gateway_tests/util.py

```diff
@@ -197,3 +201,297 @@ def caget_many_from_host(hostname, *pvnames):
         "hostname": hostname,
         "pvs": results,
     }
+
+
+class AccessBehavior(enum.IntEnum):
+    DISCONNECTED = 0
+    READ = 1
+    WRITE = 2
+
+
+def interpret_access(access):
+    try:
+        return AccessBehavior(access)
+    except ValueError:
+        pass
+    try:
+        return AccessBehavior[access]
+    except KeyError:
+        pass
+    if access.startswith('WRITE'):
+        return AccessBehavior.WRITE
+    raise ValueError(f'Could not interpret {access} as an AccessBehavior.')
+
+
+def promote_access(current, new):
+    current = interpret_access(current)
+    new = interpret_access(new)
+    if new > current:
+        return new
+    else:
+        return current
+
+
+def demote_access(current, new):
+    current = interpret_access(current)
+    new = interpret_access(new)
+    if new < current:
+        return new
+    else:
+        return current
+
+
+def correct_gateway_pvinfo(config: PCDSConfiguration, pvinfo: PVInfo,
+                           hostname: str) -> PVInfo:
+    """
+    Determine what the gateway should have given based on what the IOC gave us.
+
+    This should be mostly the same, but potentially with a modified access field
+    or possibly disconnected.
+
+    Parameters
+    ----------
+    config : PCDSConfiguration
+        All of the configuration info about the deployed env.
+    pvinfo : PVInfo
+        The pv info retrieved from the IOC.
+    hostname : str
+        The host we are testing on (or spoofing for)
+
+    Returns
+    -------
+    gwinfo : PVInfo
+        The pv info we should see from the gateway if all is well. Note that this
+        will not fill in the pvinfo address field for the gateway.
+    """
+    # If we timed out, the gateway should also time out
+    if pvinfo.error == 'timeout':
+        return pvinfo
+
+    # Otherwise, we need to determine which access rules apply.
+    # First, we need to know which subnet the PV is on.
+    subnet = get_pcds_subnet(pvinfo.address[0])
+
+    # With the subnet, we can determine which gateway rules are relevant.
+    # Only the lowest down in each pvlist file is relevant.
+    filenames = set()
+    deny_matches = {}
+    allow_matches = {}
+    for match in config.gateway_config.get_matches(pvinfo.name).matches:
+        if not os.path.basename(match.filename).startswith(subnet):
+            continue
+        filenames.add(match.filename)
+        if match.rule.command == 'DENY':
+            deny_matches[match.filename] = match
+        elif 'ALLOW' in match.rule.command:
+            allow_matches[match.filename] = match
+        elif 'DENY FROM' in match.rule.command:
+            if match.rule.command == f'DENY FROM {hostname}':
```

Comment:
I think this will be a bit different:

https://github.com/klauer/whatrecord/blob/1893ea2d0badb15f11f201b625fb385b738108c3/whatrecord/tests/test_gateway.py#L48-L57

Reply:
thanks, I'll adjust appropriately